ruby-gem/lib/calabash-android/monkey_helpers.rb

```diff
@@ -0,0 +1,91 @@
+require 'calabash-android/env'
+
+module Calabash
+  module Android
+    module MonkeyHelpers
+      include Calabash::Android::EnvironmentHelpers
+
+      MAX_RETRIES = 10
+      @@monkey_port = nil
+      @@monkey_pid = nil
+
+      def monkey_move_from(from_x, from_y, to_x, to_y, args={})
+        @@monkey_port = start_monkey
+        monkey_touch(:down, from_x, from_y)
+        sleep(args.fetch(:hold_time))
+
+        x_delta = to_x - from_x
+        y_delta = to_y - from_y
+        steps = args.fetch(:steps)
+
+        steps.times do |i|
+          move_x = (from_x + ((i+1) * (x_delta.to_f / steps))).to_i
+          move_y = (from_y + ((i+1) * (y_delta.to_f / steps))).to_i
+          monkey_touch(:move, move_x, move_y)
+        end
+
+        sleep(args.fetch(:hang_time))
+        monkey_touch(:up, to_x, to_y)
+      end
+
+      def start_monkey
+        unless xamarin_test_cloud? #monkey processes are automatically killed in XTC
+          kill_existing_monkey_processes
+        end
+
+        MAX_RETRIES.times do
+          port = rand((1024..65535))
+
+          monkey_starter_thread = Thread.new do
+            Thread.current[:output]= `#{adb_command} shell monkey --port #{port}`
```

Comment:
I think this is potentially racey code. Ruby locals are shared by the blocks of each thread. What might happen if the call is slow but fails. See also this article:  https://bibwild.wordpress.com/2015/01/15/ruby-threads-gotcha-with-local-vars-and-shared-state/

I think this will work in practice most of the time - but if we need to move fast, we should consider merging and then looking at a more robust solution.